use of unresolved module or unlinked crate `wasm_bindgen_futures`

Following #5049 I tried to remove the dependency on `wasm-bindgen-futures` and add the new optional `"futures"` feature on the `js-sys` crate. However, the `#[wasm_bindgen]` macro is now producing an error when building:

```toml
[package]
name = "microgpt-rs"
version = "0.1.0"
edition = "2024"

[dependencies]
rand = { version = "0.10.0" }
rand_distr = "0.6.0"
getrandom = { version = "0.4.2", features = ["wasm_js"] }
wasm-bindgen = { version = "0.2.117" }
# wasm-bindgen-futures = "0.4.67"
web-sys = { version = "0.3.94", features = ["console", 'Response', 'Window'] }
js-sys = { version = "0.3.94", features = ["futures"] }

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
reqwest = { version = "0.13.2", features = ["blocking"] }

[lib]
crate-type = ["cdylib", "rlib"]
```

```rust
...
#[wasm_bindgen]
pub async fn main_() -> Result<(), JsValue> {
    let mut rng = rand::rng();
    // Let there be a Dataset `docs`: list[str] of documents (e.g. a list of names)
    let window: web_sys::Window = web_sys::window().unwrap();
    // let resp_value: JsValue = JsFuture::from(window.fetch_with_str("https://raw.githubusercontent.com/karpathy/makemore/988aa59/names.txt")).await?;
    // let names = JsFuture::from(resp.text()?).await?.as_string().unwrap();
    let resp_value = window
        .fetch_with_str("https://raw.githubusercontent.com/karpathy/makemore/988aa59/names.txt")
        .await?;
    let resp: Response = resp_value.dyn_into().unwrap();
    let names: String = resp.text()?.await?.as_string().unwrap();
    ...
}
```

```rust
➜ microgpt git:(master) ✗ wasm-pack build --target web --no-typescript --no-pack
...
error[E0433]: failed to resolve: use of unresolved module or unlinked crate `wasm_bindgen_futures`                
   --> src/lib.rs:287:1
    |
287 | #[wasm_bindgen]
    | ^^^^^^^^^^^^^^^ use of unresolved module or unlinked crate `wasm_bindgen_futures`
    |
    = help: if you wanted to use a crate named `wasm_bindgen_futures`, use `cargo add wasm_bindgen_futures` to add it to your `Cargo.toml`
    = note: this error originates in the attribute macro `wasm_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)
...
```

Hi @guybedford just following this up after the latests `0.2.120.0` release.

Using the [updated documentation advice](https://wasm-bindgen.github.io/wasm-bindgen/reference/js-promises-and-rust-futures.html?highlight=future#using-js_sys-directly-without-wasm-bindgen-futures), and with the following `Cargo.toml`:

```toml
[package]
name = "microgpt-rs"
version = "0.1.0"
edition = "2024"

[dependencies]
rand = { version = "0.10.0" }
rand_distr = "0.6.0"
getrandom = { version = "0.4.2", features = ["wasm_js"] }
wasm-bindgen = { version = "0.2.120" }
# wasm-bindgen-futures = "0.4.67"
web-sys = { version = "0.3.97", features = ["console", 'Response', 'Window'] }
js-sys = { version = "0.3.97"}

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
reqwest = { version = "0.13.2", features = ["blocking"] }

[lib]
crate-type = ["cdylib", "rlib"]
```

I get this error

```shell
❯ RUSTFLAGS="--cfg=wasm_bindgen_use_js_sys" cargo build --target wasm32-unknown-unknown                                          
   Compiling wasm-bindgen-shared v0.2.120
   ...
error[E0433]: cannot find module or crate `wasm_bindgen_futures` in this scope
   --> src/lib.rs:288:1
    |
288 | #[wasm_bindgen]
    | ^^^^^^^^^^^^^^^ use of unresolved module or unlinked crate `wasm_bindgen_futures`
    |
    = help: if you wanted to use a crate named `wasm_bindgen_futures`, use `cargo add wasm_bindgen_futures` to add it to your `Cargo.toml`
    = note: this error originates in the attribute macro `wasm_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0433`.
error: could not compile `microgpt-rs` (lib) due to 1 previous error
```

If I re-add the `futures` feature

```toml
js-sys = { version = "0.3.97", features = ['futures'] }
```

it can't resolve that feature (it seems that's been factored out in lieu of the `RUSTFLAGS` parameter instead?).

```shell
❯ RUSTFLAGS="--cfg=wasm_bindgen_use_js_sys" cargo build --target wasm32-unknown-unknown
error: failed to select a version for `js-sys`.
    ... required by package `microgpt-rs v0.1.0 (/Users/yossifrenkel/Developer/microgpt)`
versions that meet the requirements `^0.3.97` (locked to 0.3.97) are: 0.3.97

package `microgpt-rs` depends on `js-sys` with feature `futures` but `js-sys` does not have that feature.
 available features: default, futures-core-03-stream, std, unsafe-eval


failed to select a version for `js-sys` which could resolve this conflict
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,12 @@
 
 ### Added
 
+* The js-sys futures codegen opt-in can now also be enabled via the
+  `WASM_BINDGEN_USE_JS_SYS=1` environment variable, in addition to
+  `--cfg=wasm_bindgen_use_js_sys`. This works on stable when `--target`
+  is in use, where Cargo does not propagate the cfg to host proc-macros.
+  [#5164](https://github.com/wasm-bindgen/wasm-bindgen/pull/5164)
+
 ### Changed
 
 ### Fixed

--- a/crates/macro-support/src/ast.rs
+++ b/crates/macro-support/src/ast.rs
@@ -9,7 +9,13 @@ use syn::Path;
 use wasm_bindgen_shared as shared;
 
 pub fn use_js_sys_futures() -> bool {
-    cfg!(wasm_bindgen_use_js_sys)
+    // Honor either the build-time cfg or an expansion-time environment variable.
+    // The env-var form is necessary because `cfg!(...)` is resolved when this
+    // proc-macro crate is itself compiled (on the host), and Cargo does not pass
+    // `--cfg`/`RUSTFLAGS` to host proc-macros when `--target` is used. Reading
+    // an env var at expansion time provides a stable workflow that works
+    // regardless of how the consumer configures Cargo.
+    cfg!(wasm_bindgen_use_js_sys) || std::env::var_os("WASM_BINDGEN_USE_JS_SYS").is_some()
 }
 
 /// Whether a function is a start function, and if so, whether it

--- a/guide/src/reference/js-promises-and-rust-futures.md
+++ b/guide/src/reference/js-promises-and-rust-futures.md
@@ -361,12 +361,28 @@ async fn process_numbers() -> Result<f64, JsValue> {
 ## Using `js_sys` directly without `wasm-bindgen-futures`
 
 To make `#[wasm_bindgen]` emit `js_sys::futures` directly and drop the
-`wasm-bindgen-futures` dependency, depend on `js-sys` and build with
-`--cfg=wasm_bindgen_use_js_sys` (e.g. via `RUSTFLAGS` or `.cargo/config.toml`).
+`wasm-bindgen-futures` dependency, depend on `js-sys` and enable the opt-in
+in one of two equivalent ways:
 
-A cfg is used rather than a Cargo feature so the choice stays scoped to the
-crate that opts in — Cargo features union across the dep graph, which would
-flip codegen for every `#[wasm_bindgen]` user in the build.
+* Set the `WASM_BINDGEN_USE_JS_SYS=1` environment variable when invoking
+  Cargo, or
+* Build with `--cfg=wasm_bindgen_use_js_sys` (e.g. via `RUSTFLAGS` or
+  `.cargo/config.toml`).
+
+The environment-variable form is read at macro expansion time and therefore
+works on stable Rust even when `--target` is used. The `--cfg` form is
+resolved when `wasm-bindgen-macro-support` itself is compiled on the host;
+Cargo does not propagate `RUSTFLAGS` or `target.*.rustflags` to host
+proc-macros when `--target` is set (see
+[rust-lang/cargo#3349](https://github.com/rust-lang/cargo/issues/3349)),
+so the cfg-only path requires nightly with
+`-Z host-config -Z target-applies-to-host` or a host build that does not
+use `--target`.
+
+A cfg/env opt-in is used rather than a Cargo feature so the choice stays
+scoped to the crate that opts in — Cargo features union across the dep
+graph, which would flip codegen for every `#[wasm_bindgen]` user in the
+build.
 
 ## Compatibility note
 

